File: py/tuish/fingerprint.py
"""Machine fingerprint generation for license binding."""

import getpass
import hashlib
import platform
import socket


def get_machine_fingerprint() -> str:
    """
    Generate a stable machine fingerprint for license binding.

    Uses: hostname + username + platform + architecture
    Matches TypeScript SDK implementation.
    """
    components = [
        socket.gethostname(),
        getpass.getuser(),
        _map_platform(platform.system()),
        _map_arch(platform.machine()),
    ]

    fingerprint_input = ":".join(components)
    return hashlib.sha256(fingerprint_input.encode("utf-8")).hexdigest()


def _map_platform(value: str) -> str:
    normalized = value.lower()
    if normalized == "macos":
        return "darwin"
    if normalized == "windows":
        return "win32"
    return normalized


def _map_arch(value: str) -> str:
    normalized = value.lower()
    if normalized in {"x86_64", "amd64"}:
        return "x64"
    if normalized in {"aarch64", "arm64"}:
        return "arm64"
    if normalized in {"x86", "i386", "i686"}:
        return "ia32"
    if normalized == "arm":
        return "arm"
    return normalized

File: go/fingerprint.go
package tuish

import (
	"crypto/sha256"
	"encoding/hex"
	"os"
	"os/user"
	"runtime"
	"strings"
)

// GetMachineFingerprint returns a stable machine fingerprint for license binding.
// Uses a combination of hostname, username, platform, and architecture.
func GetMachineFingerprint() string {
	var components []string

	// Hostname
	hostname, err := os.Hostname()
	if err != nil {
		hostname = "unknown"
	}
	components = append(components, hostname)

	// Username
	currentUser, err := user.Current()
	if err != nil {
		components = append(components, "unknown")
	} else {
		components = append(components, currentUser.Username)
	}

	// Platform (darwin, linux, windows)
	platform := mapPlatform(runtime.GOOS)
	components = append(components, platform)

	// Architecture (amd64, arm64, etc.)
	arch := mapArch(runtime.GOARCH)
	components = append(components, arch)

	// Join with colons and hash
	combined := strings.Join(components, ":")
	hash := sha256.Sum256([]byte(combined))

	return hex.EncodeToString(hash[:])
}

func mapPlatform(value string) string {
	switch value {
	case "macos":
		return "darwin"
	case "windows":
		return "win32"
	default:
		return strings.ToLower(value)
	}
}

func mapArch(value string) string {
	switch value {
	case "x86_64", "amd64":
		return "x64"
	case "aarch64":
		return "arm64"
	case "x86", "i386", "i686", "386":
		return "ia32"
	default:
		return strings.ToLower(value)
	}
}
